### ISSUE TITLE
Crash when changing map parts

### ISSUE THREAD

**Issue by lpechacek** [OpenOrienteering/mapper#1892]
I see this issue occasionally with my development builds. I thought that there is something wrong with my build system, but I spotted the crash with a build artifact from Pipelines yesterday. I'm filing the ticket to record my findings and collect ideas about what might be the root cause.

### Steps to reproduce
1. Open a map with many GPX tracks and more than one map part
2. Make sure that the Map parts action is in the "overflow menu" (the three dots)
3. Try to change the map part while the GPX tracks are loading in the background

### Actual behaviour
Null pointer dereference in `libQt5Widgets.so` at the moment when the map part selection should appear. The issue replicates on about 50-70% of attempts.

### Expected behaviour
No crash.

### Configuration
Mapper Version: master-v20210302.1 build in Pipelines, signed with own key
Operating System: Android 9, Samsung edition for Galaxy Tab Active2 (SM-T390)

### Developer remarks
This feels like some race. The backtraces of crashed threads do not provide insight into the sequence of actions that led to the event. The dereferenced address is 0x4, which suggests member access through a null pointer. I haven't yet studied the neighboring thread backtraces. Maybe they can give a clue about the application status.

One of the tombstones: [tombstone_03.txt](https://github.com/OpenOrienteering/mapper/files/6083358/tombstone_03.txt)
 The application tombstones are all alike. I can provide more on request.

**Comment by ollesmaps** [comment]
Might be related?

1.  Open a map with many templates
2.  It takes some time to load them all
3.  Switch Location display toggle on during the template load
4. OOM crashes

**Comment by lpechacek** [comment]
>     3. Switch Location display toggle on during the template load

It may or may not be the same issue. Location display toggle creates a new file in the storage but does not change the user interface appearance. In contrast, the map part switch creates an on-display popup with the list of parts. That said, I wouldn't expect a crash in `libQt5Widgets.so`, which is the code for the graphical user interface, in the Location display toggle case.

Regarding your particular issue, collection of the crash reports would help. I think I wrote the procedure elsewhere but repetition cannot hurt:
1) Enable Android Developer Options - https://developer.android.com/studio/debug/dev-options
2) As soon as possible after a crash, generate a bug report - tap Take bug report (Vytvořit chybové hlášení) in the developer options menu - https://developer.android.com/studio/debug/bug-report
3) After the debugging information is collected, think twice before you copy data out of the device. The bug report contains the list of installed applications, traces of recent interaction with the device at al. On the other hand, it does not contain any user data, o-map data, passwords, and not even the name of the open map file and templates. If you are fine with sharing this information with us, proceed. (Google already has similar data about you anyway. ;-) )
4) At this point, either send the report data (a few megabytes) directly to your favorite OpenOrienteering developer or send it over to your computer and filter out only the relevant application crash reports for sharing.

Thanks!

**Comment by lpechacek** [comment]
Just replicated the crash with v0.9.4.

**Comment by lpechacek** [comment]
Just figured out that the app works fine when the map parts action sits in the toolbar. The crash occurs when it's in the "overflow menu". The action renders as the icon, text, and a triangle pointing to the right at the right edge of the display. The crash occurs when I activate the map parts action with *stylus*. I wasn't able to trigger the crash with the display touch. I succeeded once with an optical mouse.

**Comment by dg0yt** [comment]
Is related to asynchronous template loading at all? Anyway, the template loading does not happen in the background (other thread) at the moment. It happens right in the main thread, blocking the event loop, and thus possibly interfering with processing of user input and UI update events.

In addition, the map part menu is updated on the hover event...
https://github.com/OpenOrienteering/mapper/blob/b86983bc5ad983324addfb2e66fd4bb167779f27/src/gui/map/map_editor.cpp#L1457-L1472

Hovering works well with mouse or stylus (tooltips!), but touch?

Definitely not equally tested for all contexts (touch or not, regular or overflow, template loading vs. later).
HTH

**Comment by dg0yt** [comment]
> 3\. Switch Location display toggle on during the template load

In particular, it adds a template during template loading. It should not fail, but it is easy to get this wrong. We should track this in a separate issue.

**Comment by lpechacek** [comment]
> Is it related to asynchronous template loading at all?

The plate loading changes the event timing so that the race can trigger. Nothing else, I'd say.

> Definitely not equally tested for all contexts (touch or not, regular or overflow, template loading vs. later).

Correct code does not have to be tested for correctness. ;-D

> HTH

It does. Now I can trigger the crash reliably with mouse and stylus. During template loading, I click the overflow button, click the Map parts action and immediately click the overflow button again so that the drop-down disappears. I guess that the hover event then gets delivered to a deleted object or something like that.

**Comment by dg0yt** [comment]
The desktop mouse pointer should create similar events like the stylus on Android. However, I cannot reproduce this issue on the desktop with the configuration which reproduced #1893.

**Comment by lpechacek** [comment]
I'm not able to replicate the crash on any other device than the Samsung tablet.

I looked deeper into Qt internals, and in my opinion, the culprit lies in the fact that the overflow action creates a popup menu with the extra actions. The map parts action, in turn, creates a popup with the map parts. When the first popup (extra actions) disappears before the second one renders on display, Qt chokes.

My conclusions so far are:
1) This might be a Qt bug in the popup menu handling.
2) It might be a fallout of the event queue blocking and then not a Qt bug by definition.
3) The issue is limited in time. It can be triggered only during template loading and not later.
4) The crash can be easily worked around by making the buttons smaller so that the map parts action is not in the overflow menu. (which is what I did)

With that, I'm tempted to close the ticket as "won't fix" for now. The issue is very difficult to analyze, triggers rarely, has an easy workaround, and there are more important issues to fix. I'll provide a few days of a grace period for alternate opinions to be voiced.
